File: backend/app/schemas/billing.py
from __future__ import annotations

from datetime import datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, ConfigDict, Field, model_validator


class PlanResponse(BaseModel):
    id: UUID
    version_id: UUID
    code: str
    display_name: str
    description: str
    version: int
    currency: str
    monthly_price_minor: int | None
    yearly_price_minor: int | None
    trial_days: int
    active: bool
    public: bool
    entitlements: dict[str, bool | int]


class BillingOverviewResponse(BaseModel):
    business_id: UUID
    subscription_id: UUID | None
    plan_id: UUID
    plan_version_id: UUID
    plan_code: str
    plan_name: str
    plan_version: int
    subscription_status: str
    access_reason: str
    billing_interval: str
    current_period_start: datetime
    current_period_end: datetime
    trial_started_at: datetime | None
    trial_ends_at: datetime | None
    cancel_at_period_end: bool
    entitlements: dict[str, bool | int]
    provider_configured: bool = False


class UsageResponse(BaseModel):
    period_start: datetime
    period_end: datetime
    usage: dict[str, int]
    limits: dict[str, int]
    remaining: dict[str, int]
    informational: dict[str, int]


class PlanChangeIntentRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    plan_code: str = Field(min_length=1, max_length=48, pattern=r"^[a-z][a-z0-9_]*$")
    billing_interval: Literal["month", "year"]


class PlanChangeIntentResponse(BaseModel):
    status: Literal["provider_unavailable", "blocked", "checkout_ready"]
    message: str
    blockers: list[dict[str, int | str]] = []
    checkout_url: str | None = None


class CancellationRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    reason: str = Field(min_length=3, max_length=500)


class SubscriptionMutationResponse(BaseModel):
    status: str
    cancel_at_period_end: bool
    current_period_end: datetime


class AdminPlanAvailabilityRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    active: bool
    public: bool
    reason: str = Field(min_length=3, max_length=500)


class AdminPlanVersionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    currency: str = Field(default="USD", min_length=3, max_length=3, pattern=r"^[A-Z]{3}$")
    monthly_price_minor: int | None = Field(default=None, ge=0)
    yearly_price_minor: int | None = Field(default=None, ge=0)
    entitlements: dict[str, bool | int]
    reason: str = Field(min_length=3, max_length=500)


class AdminSubscriptionAssignRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    plan_code: str = Field(min_length=1, max_length=48, pattern=r"^[a-z][a-z0-9_]*$")
    billing_interval: Literal["month", "year"] = "month"
    trial_days: int = Field(default=0, ge=0, le=365)
    reason: str = Field(min_length=3, max_length=500)


class AdminTrialExtensionRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    days: int = Field(ge=1, le=365)
    reason: str = Field(min_length=3, max_length=500)


class AdminSubscriptionStatusRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    status: Literal["active", "suspended"]
    reason: str = Field(min_length=3, max_length=500)


class AdminEntitlementOverrideRequest(BaseModel):
    model_config = ConfigDict(extra="forbid")
    entitlement_key: str = Field(min_length=1, max_length=64, pattern=r"^[a-z][a-z0-9_]*$")
    boolean_value: bool | None = None
    integer_value: int | None = Field(default=None, ge=0)
    active: bool = True
    expires_at: datetime | None = None
    reason: str = Field(min_length=3, max_length=500)

    @model_validator(mode="after")
    def exactly_one_value(self) -> "AdminEntitlementOverrideRequest":
        if (self.boolean_value is None) == (self.integer_value is None):
            raise ValueError("Exactly one typed entitlement value is required")
        return self


class AdminSubscriptionItem(BaseModel):
    business_id: UUID
    subscription_id: UUID
    plan_code: str
    plan_name: str
    status: str
    billing_interval: str
    current_period_end: datetime
    cancel_at_period_end: bool


class AdminSubscriptionPage(BaseModel):
    items: list[AdminSubscriptionItem]
    page: int
    page_size: int
    total: int


class AdminSubscriptionEventItem(BaseModel):
    id: UUID
    subscription_id: UUID
    business_id: UUID
    event_type: str
    actor_user_id: UUID | None
    from_status: str | None
    to_status: str | None
    from_plan_version_id: UUID | None
    to_plan_version_id: UUID | None
    reason: str | None
    created_at: datetime


class AdminSubscriptionEventPage(BaseModel):
    items: list[AdminSubscriptionEventItem]
    page: int
    page_size: int
    total: int


class AdminBillingAuditItem(BaseModel):
    id: UUID
    business_id: UUID | None
    actor_user_id: UUID | None
    event_type: str
    target_type: str
    target_id: UUID | None
    before_state: dict[str, object]
    after_state: dict[str, object]
    reason: str
    created_at: datetime


class AdminBillingAuditPage(BaseModel):
    items: list[AdminBillingAuditItem]
    page: int
    page_size: int
    total: int


class AdminBillingMetrics(BaseModel):
    subscriptions_by_status: dict[str, int]
    subscriptions_by_plan: dict[str, int]
    businesses_without_subscription: int
    note: str = "Revenue metrics are unavailable because no payment provider is configured."

File: artifacts/ai-business-os/src/services/billing.ts
import { apiClient } from "./api-client.ts";

export type BillingEntitlements = Record<string, boolean | number>;

export type BillingOverview = {
  business_id: string;
  subscription_id: string | null;
  plan_id: string;
  plan_version_id: string;
  plan_code: string;
  plan_name: string;
  plan_version: number;
  subscription_status: string;
  access_reason: string;
  billing_interval: "month" | "year";
  current_period_start: string;
  current_period_end: string;
  trial_started_at: string | null;
  trial_ends_at: string | null;
  cancel_at_period_end: boolean;
  entitlements: BillingEntitlements;
  provider_configured: boolean;
  test_plan_activation_enabled: boolean;
};

export type BillingUsage = {
  period_start: string;
  period_end: string;
  usage: Record<string, number>;
  limits: Record<string, number>;
  remaining: Record<string, number>;
  informational: Record<string, number>;
};

export type BillingPlan = {
  id: string;
  version_id: string;
  code: string;
  display_name: string;
  description: string;
  version: number;
  currency: string;
  monthly_price_minor: number | null;
  yearly_price_minor: number | null;
  trial_days: number;
  active: boolean;
  public: boolean;
  entitlements: BillingEntitlements;
};

export type PlanChangeIntent = {
  status:
    | "provider_unavailable"
    | "blocked"
    | "checkout_ready"
    | "test_activated";
  message: string;
  blockers: Array<{
    entitlement_key: string;
    current: number;
    target_limit: number;
  }>;
  checkout_url: string | null;
  billing: BillingOverview | null;
};

export function isCurrentBillingResponse(
  requestedBusinessId: string,
  responseBusinessId: string,
  requestVersion: number,
  currentVersion: number,
  activeBusinessId: string,
) {
  return (
    requestedBusinessId === responseBusinessId &&
    requestedBusinessId === activeBusinessId &&
    requestVersion === currentVersion
  );
}

export function shouldRequestPlanChange(
  currentPlanCode: string,
  targetPlanCode: string,
) {
  return currentPlanCode !== targetPlanCode;
}

export function billingPlanActionLabel(
  current: boolean,
  planCode: string,
  displayName: string,
  providerConfigured: boolean,
  testPlanActivationEnabled: boolean,
) {
  if (current) return "Selected";
  if (testPlanActivationEnabled && planCode !== "free") {
    return `Activate ${displayName}`;
  }
  return providerConfigured ? "Choose plan" : "Check availability";
}

function root(businessId: string) {
  return `/api/v1/businesses/${encodeURIComponent(businessId)}/billing`;
}

export function createBillingApi(client = apiClient) {
  return {
  overview: (businessId: string, signal?: AbortSignal) =>
    client.request<BillingOverview>(root(businessId), { signal }),
  usage: (businessId: string, signal?: AbortSignal) =>
    client.request<BillingUsage>(`${root(businessId)}/usage`, { signal }),
  plans: (businessId: string, signal?: AbortSignal) =>
    client.request<BillingPlan[]>(`${root(businessId)}/plans`, { signal }),
  changeIntent: (
    businessId: string,
    planCode: string,
    billingInterval: "month" | "year",
  ) =>
    client.request<PlanChangeIntent>(`${root(businessId)}/change-intent`, {
      method: "POST",
      json: { plan_code: planCode, billing_interval: billingInterval },
    }),
  cancel: (businessId: string, reason: string) =>
    client.request<{
      status: string;
      cancel_at_period_end: boolean;
      current_period_end: string;
    }>(`${root(businessId)}/cancel`, {
      method: "POST",
      json: { reason },
    }),
  reactivate: (businessId: string) =>
    client.request<{
      status: string;
      cancel_at_period_end: boolean;
      current_period_end: string;
    }>(`${root(businessId)}/reactivate`, { method: "POST" }),
  };
}

export const billingApi = createBillingApi();
